src/rdkafka_msg.c

```diff
@@ -543,33 +543,29 @@ int32_t rd_kafka_msg_partitioner_consistent_random (const rd_kafka_topic_t *rkt,
                                                  msg_opaque);
 }
 
-int32_t rd_kafka_msg_partitioner_murmur2_consistent (const rd_kafka_topic_t *rkt,
-					 const void *key, size_t keylen,
-					 int32_t partition_cnt,
-					 void *rkt_opaque,
-					 void *msg_opaque) {
-	return rd_murmur2(key, keylen) % partition_cnt;
+int32_t
+rd_kafka_msg_partitioner_murmur2 (const rd_kafka_topic_t *rkt,
+                                  const void *key, size_t keylen,
+                                  int32_t partition_cnt,
+                                  void *rkt_opaque,
+                                  void *msg_opaque) {
+        return rd_murmur2(key, keylen) % partition_cnt;
 }
 
 int32_t rd_kafka_msg_partitioner_murmur2_random (const rd_kafka_topic_t *rkt,
-					 const void *key, size_t keylen,
-					 int32_t partition_cnt,
-					 void *rkt_opaque,
-					 void *msg_opaque) {
-	if (!key)
-		return rd_kafka_msg_partitioner_random(rkt,
-						       key,
-						       keylen,
-						       partition_cnt,
-						       rkt_opaque,
-						       msg_opaque);
-	else
-		return rd_kafka_msg_partitioner_murmur2_consistent(rkt,
-								   key,
-								   keylen,
-								   partition_cnt,
-								   rkt_opaque,
-								   msg_opaque);
+                                                 const void *key, size_t keylen,
+                                                 int32_t partition_cnt,
+                                                 void *rkt_opaque,
+                                                 void *msg_opaque) {
+        if (!key)
```

Comment:
minor: in `rd_kafka_msg_partitioner_consistent_random` the test is keylen == 0. I assume you organize for this to be the case in NULL case as well but you could consider explicitly doing a !key check there to make it easy to read.

Reply:
but for this method, yep I verified this logic matches the java default partitioner.

Reply:
This is a legacy mistake, the consistent_random partitioner treats NULL and empty keys the same way, but only NULL keys should be randomized.

I'm however reluctant to change this since it will change the behaviour of existing applications.

The murmur2_random partitioner follows the Java default partitioner and only checks for NULL.